Fill.py

```diff
@@ -452,17 +454,106 @@ def distribute_early_items(multiworld: MultiWorld,
+
+    fill_locations = [location for location in original_order if not location.item]
+
     return fill_locations, itempool
 
 
+def balanced_shuffle(multiworld: MultiWorld, fill_locations: list[Location], itempool: list[Item]) -> list[Location]:
+    balancing_factor = 0.5  # This would be a setting instead. Acts as a percentage. 0.0 is min, 1.0 is max.
+
+    # First, we shuffle the location pool.
+    multiworld.random.shuffle(fill_locations)
+
+    # If balancing factor is 0, don't do any more unnecessary work.
+    if balancing_factor == 0.0:
+        return fill_locations
+
+    # Balancing only has an effect if there is progression
+    amount_of_progression = sum(1 for item in itempool if item.advancement)
+    if not amount_of_progression:
+        return fill_locations
+
+    # If balancing factor is not 0, we split up the locations list by players.
+    locations_per_player = collections.defaultdict(list)
+    for location in fill_locations:
+        locations_per_player[location.player].append(location)
+
+    # shuffle the player order in the dict so that no player gets an inherent advantage.
+    players_with_locations = list(locations_per_player)
+    multiworld.random.shuffle(players_with_locations)
+    locations_per_player = {player: locations_per_player[player] for player in players_with_locations}
+
+    # Grab some important values for later
+    fill_location_counts = {player: len(locations) for player, locations in locations_per_player.items()}
+    total_fill_locations = sum(fill_location_counts.values())
+
+    # Now, the actual balancing begins.
+    # We will have two sets of weights.
+
+    # The first set of weights will be the number of progression items that are expected to end up in each world,
+    # assuming full randomness.
+    # Using this set of weights to shuffle the list will just, on average, just result in a random shuffle.
+
+    expected_progression_counts_if_random = {
+        player: fill_location_count / total_fill_locations * amount_of_progression
+        for player, fill_location_count in fill_location_counts.items()
+    }
+
+    # The second set of weights is the number of progression items that would end up in each world if we tried to place
+    # an equal amount of progression into each world.
+    # This could be just one shared value for all worlds (number of progression items divided by number of slots), but
+    # we have to make sure that we don't give any game more locations than it can hold.
+
+    balanced_progression_counts = {player: 0 for player in fill_location_counts}
+
+    progression_to_distribute = amount_of_progression
+    while True:
+        for player, max_count in fill_location_counts.items():
+            if balanced_progression_counts[player] == max_count:
+                continue
+            balanced_progression_counts[player] += 1
+            progression_to_distribute -= 1
+            if progression_to_distribute == 0:
+                break
+        else:
```

Comment:
```suggestion
        else:  # have not distributed all progression yet
```
just making it easier to understand